lib/tree.rb

```diff
@@ -2,59 +2,124 @@ class TreeNode
 
-   def initialize(key, val)
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
+
   def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: o(n)
+  # Space Complexity: o(n)
   def add(key, value)
-    raise NotImplementedError
+    current_node = @root
+
+    if current_node == nil
+      current_node = TreeNode.new(key, value)
+      @root = current_node
+    end
+
+    while nil != current_node
+      if key < current_node.key && current_node.left == nil
+        current_node.left = TreeNode.new(key, value)
+      elsif key > current_node.key && current_node.right == nil
+        current_node.right = TreeNode.new(key, value)
+      elsif key > current_node.key
+        current_node = current_node.right
+      elsif key < current_node.key
+        current_node = current_node.left
+      else
+        return
+      end
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: o(n)
+  # Space Complexity: 0(1)
   def find(key)
-    raise NotImplementedError
+    return nil if @root.nil?
+    current_node = @root
+
+    return current_node.value if key == current_node.key
+
+    while current_node != key
+      if key == current_node.key
+        return current_node.value
+      elsif key > current_node.key
+        current_node = current_node.right
+      elsif key < current_node.key
+        current_node = current_node.left
+      end
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+# Time Complexity: O(n)
+# Space Complexity: O(1)
+  # left - Root - Rigth
   def inorder
-    raise NotImplementedError
+    inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(node = @root, list)
+    return list if node.nil?
+    inorder_helper(node.left, list)
+    list << { :key => node.key, :value => node.value }
+    inorder_helper(node.right, list)
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  # Root - left - Rigth
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(node = @root, list)
+    return list if node.nil?
+    list << { :key => node.key, :value => node.value }
+    preorder_helper(node.left, list)
+    preorder_helper(node.right, list)
+
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  # left - Rigth - Root
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root, [])
+  end
+
+  def postorder_helper(node = @root, list)
+    return list if node.nil?
+    postorder_helper(node.left, list)
+    postorder_helper(node.right, list)
+    list << { :key => node.key, :value => node.value }
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
   def height
```

Comment:
👍 See above in space/time.  The space complexity is either O(log n) for a balanced tree or O(n) for an unbalanced one.